index/scorch/snapshot_index.go

```diff
@@ -492,124 +491,115 @@ func (i *IndexSnapshot) DocumentVisitFieldTerms(id index.IndexInternalID,
+	ss := i.segment[segmentIndex]
 
-		go func() {
-			defer close(errCh)
-			err := ss.cachedDocs.prepareFields(dvPendingFields, ss)
-			if err != nil {
-				errCh <- err
-			}
-		}()
+	var vFields []string // fields that are visitable via the segment
 
-		// visit the requested persisted dv while the cache preparation in progress
-		dvs, err = zaps.VisitDocumentFieldTerms(localDocNum, fields, visitor, dvs)
+	ssv, ssvOk := ss.segment.(segment.DocumentFieldTermVisitable)
+	if ssvOk && ssv != nil {
+		vFields, err = ssv.VisitableDocValueFields()
 		if err != nil {
-			return nil, err
+			return nil, nil, err
 		}
-
-		// err out if fieldCache preparation failed
-		err = <-errCh
-		if err != nil {
-			return nil, err
-		}
-
-		visitDocumentFieldCacheTerms(localDocNum, dvPendingFields, ss, visitor)
-		return dvs, nil
 	}
 
-	return dvs, prepareCacheVisitDocumentFieldTerms(localDocNum, fields, ss, visitor)
-}
+	var errCh chan error
 
-func prepareCacheVisitDocumentFieldTerms(localDocNum uint64, fields []string,
-	ss *SegmentSnapshot, visitor index.DocumentFieldTermVisitor) error {
-	err := ss.cachedDocs.prepareFields(fields, ss)
-	if err != nil {
-		return err
-	}
-
-	visitDocumentFieldCacheTerms(localDocNum, fields, ss, visitor)
-	return nil
-}
+	// cFields represents the fields that we'll need from the
+	// cachedDocs, and might be optionally be provided by the caller,
+	// if the caller happens to know we're on the same segmentIndex
+	// from a previous invocation
+	if cFields == nil {
+		cFields = subtractStrings(fields, vFields)
 
-func visitDocumentFieldCacheTerms(localDocNum uint64, fields []string,
-	ss *SegmentSnapshot, visitor index.DocumentFieldTermVisitor) {
+		if !ss.cachedDocs.hasFields(cFields) {
+			errCh = make(chan error, 1)
 
-	for _, field := range fields {
-		if cachedFieldDocs, exists := ss.cachedDocs.cache[field]; exists {
-			if tlist, exists := cachedFieldDocs.docs[localDocNum]; exists {
-				for {
-					i := bytes.Index(tlist, TermSeparatorSplitSlice)
-					if i < 0 {
-						break
-					}
-					visitor(field, tlist[0:i])
-					tlist = tlist[i+1:]
+			go func() {
+				err := ss.cachedDocs.prepareFields(cFields, ss)
+				if err != nil {
+					errCh <- err
 				}
-			}
+				close(errCh)
+			}()
 		}
 	}
 
-}
-
-func extractDvPendingFields(requestedFields, persistedFields []string) []string {
-	removeMap := make(map[string]struct{}, len(persistedFields))
-	for _, str := range persistedFields {
-		removeMap[str] = struct{}{}
+	if ssvOk && ssv != nil && len(vFields) > 0 {
+		dvs, err = ssv.VisitDocumentFieldTerms(localDocNum, fields, visitor, dvs)
+		if err != nil {
+			return nil, nil, err
+		}
 	}
 
-	rv := make([]string, 0, len(requestedFields))
-	for _, s := range requestedFields {
-		if _, ok := removeMap[s]; !ok {
-			rv = append(rv, s)
+	if errCh != nil {
+		err = <-errCh
+		if err != nil {
+			return nil, nil, err
 		}
 	}
-	return rv
+
+	ss.cachedDocs.visitDoc(localDocNum, cFields, visitor)
```

Comment:
one thought, by calling this every time even when it may not be needed, aren't performing the lock/unlock on every single hit?  can it be conditional?

Looks pretty good to me otherwise 💯 :)

Reply:
thanks -- added a change to the PR to reflect that